### PR TITLE
daemon: patch CNP status to avoid endless retries on update status

### PR DESCRIPTION
**Pull request by aanm**
As update status allows to be patched directly we can take advantage of
this behavior to update the CNP node status without failing for every
node that wants to update its status using UpdateStatus function from
the k8s client-go

Signed-off-by: André Martins <andre@cilium.io>

~Review per commit and **read the commit message of the last commit** since it is important~

This PR will not affect k8s <= 1.12 as the patch functionality in those versions is limited.

---
This change is [Reviewable](https://reviewable.io/reviews/cilium/cilium/7140)

### CODE REVIEW

**Comment by aanm** [comment]
test-me-please

**Comment by aanm** [comment]
test-me-please

**Comment by aanm** [comment]
test-me-please

**Comment by aanm** [comment]
test-me-please

**Comment by aanm** [comment]
test-me-please

**Comment by aanm** [comment]
test-me-please

**Comment by aanm** [comment]
test-me-please

File: daemon/k8s_watcher.go [hunk 1]
```diff
@@ -1306,13 +1329,61 @@ func updateCNPNodeStatus(ciliumNPClient clientset.Interface, cnp *cilium_v2.Cili
...
+				},
+			},
+		}
+
+		createStatusAndNodePatchJSON, err := json.Marshal(createStatusAndNodePatch)
+		if err != nil {
+			return err
+		}
+
+		_, err2 = ciliumNPClient.CiliumV2().CiliumNetworkPolicies(ns).Patch(cnp.GetName(), types.JSONPatchType, createStatusAndNodePatchJSON, "status")
```
Comment (ianvernon) [inline daemon/k8s_watcher.go]: Please use more descriptive error names than `err` and `err2` that are more evocative of what types of errors they are. This makes the code more readable. In the past CNP status update code, a similar convention to naming errors was used and it was quite difficult to read the code.

File: daemon/k8s_watcher.go [hunk 2]
```diff
@@ -1306,13 +1329,61 @@ func updateCNPNodeStatus(ciliumNPClient clientset.Interface, cnp *cilium_v2.Cili
...
+		// and only one node will succeed. This can be moved to the
+		// cilium-operator which will create the path for all nodes. However
+		// performance tests have shown that performing 2 API calls to
+		// kube-apiserver for 1000 nodes versus performing 1 API call, where
+		// one of the nodes would "create" the `/status` path before all other
+		// nodes tried to replace their own status resulted in a gain of X %
+		createStatusAndNodePatch := []jsonPatch{
+			{
+				OP:    "test",
+				Path:  "/status",
```
Comment (ianvernon) [inline daemon/k8s_watcher.go:1351]: This is testing for if the entire `/status` field is `nil`, and if so, we `add` initialize the status for the CNP with the status of this node filled. Are we guaranteed that no overwrites of the  Nodes map will occur across nodes with this approach?

File: daemon/k8s_watcher.go [hunk 3]
```diff
@@ -1306,13 +1329,61 @@ func updateCNPNodeStatus(ciliumNPClient clientset.Interface, cnp *cilium_v2.Cili
...
+		}
+
+		createStatusAndNodePatchJSON, err := json.Marshal(createStatusAndNodePatch)
+		if err != nil {
+			return err
+		}
+
+		_, err2 = ciliumNPClient.CiliumV2().CiliumNetworkPolicies(ns).Patch(cnp.GetName(), types.JSONPatchType, createStatusAndNodePatchJSON, "status")
+		if err2 != nil && k8sErrors.IsInvalid(err2) {
+			// If it fails it means the test from the previous patch failed
```
Comment (ianvernon) [inline daemon/k8s_watcher.go:1373]: To clarify - `/status` would be `nil` when no status has been created for the given CNP (yet). This would occur when the first node to update the status within the CNP tries to update its status, correct?

Can you clarify / confirm my parsing of the following?

If multiple nodes try to update the CNP NodeStatus all at the same time, then the `test` patch would fail for all of them.  If this is the case, then for all, we would `replace` here, but this is okay / safe to do because we `replace` at a unique path (`/status/nodes/<nodename>`) which will not be overwritten from other nodes doing the `replace` operation. This is because the `replace` is only for this given node. Is that correct?

File: daemon/k8s_watcher.go [hunk 4]
```diff
@@ -35,7 +35,6 @@ import (
 	"github.com/cilium/cilium/pkg/k8s/apis/cilium.io"
```
Comment (ianvernon) [inline daemon/k8s_watcher.go]: Nit - important typo in commit message - <1.10 should be <1.11

Comment (ianvernon) [inline daemon/k8s_watcher.go]: Will the error example in your commit be logged for every single CNP NodeStatus update? If this is the case, this may alarm some users since there will be a lot of errors on the K8s apiserver :(

File: daemon/k8s_watcher.go [hunk 5]
```diff
@@ -1383,8 +1316,35 @@ func updateCNPNodeStatus(ciliumNPClient clientset.Interface, cnp *cilium_v2.Cili
 			_, err2 = ciliumNPClient.CiliumV2().CiliumNetworkPolicies(ns).Patch(cnp.GetName(), types.JSONPatchType, createStatusAndNodePatchJSON, "status")
 		}
 	default:
-		cnp.SetPolicyStatus(nodeName, cnpns)
-		_, err2 = ciliumNPClient.CiliumV2().CiliumNetworkPolicies(ns).Update(cnp)
+		// During testing it was noticed kube-apiserver got panics when this
+		// patch was performed, for this reason we will not use node PatchJson.
+		_, err2 = ciliumNPClient.CiliumV2().CiliumNetworkPolicies(ns).Patch(cnp.GetName(), types.JSONPatchType, createStatusAndNodePatchJSON)
```
Comment (ianvernon) [inline daemon/k8s_watcher.go]: Same here, please do not use `err` and `err2` as names. Please use separate error names for the different errors generated, because it will make the code a lot more readable.

File: daemon/k8s_watcher.go [hunk 6]
```diff
@@ -1383,8 +1316,35 @@ func updateCNPNodeStatus(ciliumNPClient clientset.Interface, cnp *cilium_v2.Cili
...
+		if err2 != nil {
+			// If it fails it might mean the `/status` path exists.
+			// Before you ask, we can't blindly add this node status to the path
+			// `/status` because it will delete all other nodes.
+			createStatusAndNodePatch[0].Value = cilium_v2.CiliumNetworkPolicyStatus{}
```
Comment (ianvernon) [inline daemon/k8s_watcher.go]: Add a comment to indicate that this updates the `test` portion of the `createStatusAndNodePatch` to test if the NodeStatus is empty instead of `nil`.

**Review by ianvernon — CHANGES_REQUESTED** [review]
This is a great change overall! Some minor comments around naming / clarification. It took me a while to grasp what was going on with the nitty-gritty of the patch statements (this may be around my lack of understanding around patch semantics), but I think more detail there would be quite helpful for code-maintainability purposes.

File: daemon/k8s_watcher.go (hunk 2 above)
Comment (aanm) [reply daemon/k8s_watcher.go:1351]: Yes, if `/status` is not nil then it will immediately fail.

File: daemon/k8s_watcher.go (hunk 3 above)
Comment (aanm) [reply daemon/k8s_watcher.go:1373]: correct

File: daemon/k8s_watcher.go (hunk 4 above)
Comment (aanm) [reply daemon/k8s_watcher.go]: Yes

**Comment by aanm** [comment]
test-me-please

**Comment by aanm** [comment]
I've dropped the last commit as I've noticed that kube-apiserver for k8s 1.9 was getting panics.

**Comment by aanm** [comment]
test-me-please

File: daemon/k8s_watcher_test.go [hunk 7]
```diff
@@ -1097,3 +1115,570 @@ func (ds *DaemonSuite) Test_addCiliumNetworkPolicyV2(c *C) {
...
 	}
 }
+
+type K8sSuite struct{}
+
+var _ = Suite(&K8sSuite{})
+
+func (k *K8sSuite) SetUpSuite(c *C) {
+	if true {
+		logging.DefaultLogger.SetLevel(logrus.PanicLevel)
```
Comment (ianvernon) [inline daemon/k8s_watcher_test.go:1125]: This may set the logging level for the whole test suite. This isn't necessarily a problem, as I think reducing the verbosity of unit test output is a good thing.

**Review by ianvernon — APPROVED** [review]

**Comment by aanm** [comment]
test-missed-k8s

**Comment by aanm** [comment]
test-missed-k8s

**Comment by aanm** [comment]
test-me-please

**Comment by tgraf** [comment]
Only test failure is `BDD-Test-k8s-1.11-and-1.12 / k8s-1.11.K8sFQDNTest Restart Cilium validate that FQDN is still working`

File: daemon/k8s_watcher.go [hunk 8]
```diff
@@ -1185,14 +1182,22 @@ func (d *Daemon) addCiliumNetworkPolicyV2(ciliumV2Store cache.Store, cnp *cilium
...
-				return cnpNodeStatusController(ciliumV2Store, cnp, rev, scopedLog, policyImportErr)
+				return cnpNodeStatusController(ciliumNPClient, ciliumV2Store, cnp, rev, scopedLog, policyImportErr, node.GetName())
 			},
 		},
 	)
 	return policyImportErr
 }
 
-func cnpNodeStatusController(ciliumV2Store cache.Store, cnp *cilium_v2.CiliumNetworkPolicy, rev uint64, logger *logrus.Entry, policyImportErr error) error {
+func cnpNodeStatusController(
```
Comment (tgraf) [inline daemon/k8s_watcher.go:1201]: We should try to avoid formatting changes in feature and bug commits.

File: daemon/k8s_watcher.go [hunk 9]
```diff
@@ -1307,16 +1330,72 @@ func updateCNPNodeStatus(ciliumNPClient clientset.Interface, cnp *cilium_v2.Cili
...
+		// This is a JSON Patch used to create the `/status/nodes` field in the
+		// CNP. If we don't create, replacing the status for this node will
+		// fail as the path does not exist.
+		// Worst case scenario is that all nodes try to perform this operation
+		// and only one node will succeed. This can be moved to the
+		// cilium-operator which will create the path for all nodes. However
+		// performance tests have shown that performing 2 API calls to
+		// kube-apiserver for 1000 nodes versus performing 1 API call, where
+		// one of the nodes would "create" the `/status` path before all other
+		// nodes tried to replace their own status resulted in a gain of X %
```
Comment (tgraf) [inline daemon/k8s_watcher.go]: What is X? :-)

File: daemon/k8s_watcher.go [hunk 10]
```diff
@@ -1307,16 +1330,72 @@ func updateCNPNodeStatus(ciliumNPClient clientset.Interface, cnp *cilium_v2.Cili
...
+		createStatusAndNodePatch := []jsonPatch{
+			{
+				OP:    "test",
```
Comment (tgraf) [inline daemon/k8s_watcher.go:1350]: What does test do?

**Review by tgraf — CHANGES_REQUESTED** [review]

File: daemon/k8s_watcher.go (hunk 8 above)
Comment (aanm) [reply daemon/k8s_watcher.go:1201]: it's not formatting, lots of new fields were added.

File: daemon/k8s_watcher.go (hunk 10 above)
Comment (aanm) [reply daemon/k8s_watcher.go:1350]: https://tools.ietf.org/html/rfc6902#section-4.6

File: daemon/k8s_watcher.go (hunk 9 above)
Comment (aanm) [reply daemon/k8s_watcher.go]: Fixed

**Comment by aanm** [comment]
test-me-please

**Review by tgraf — APPROVED** [review]

**Comment by tgraf** [comment]
` k8s-1.13.K8sFQDNTest Restart Cilium validate that FQDN is still working`

**Comment by tgraf** [comment]
test-me-please

**Comment by aanm** [comment]
Hit https://github.com/cilium/cilium/issues/7282

**Comment by aanm** [comment]
test-missed-k8s
